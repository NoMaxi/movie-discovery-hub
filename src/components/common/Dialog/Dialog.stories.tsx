import type { Meta, StoryObj } from "@storybook/react";
import { fn } from "@storybook/test";
import { mockMovieInfo } from "@/mocks/MovieData";
import { MovieForm } from "@/components/MovieForm/MovieForm";
import { Dialog } from "./Dialog";

const meta = {
    title: "Common/Dialog",
    component: Dialog,
    argTypes: {
        title: {
            control: "text",
            description: "The title content for the dialog header. Can be a string or JSX.",
        },
        children: {
            control: false,
            description: "The main content of the dialog body.",
        },
        onClose: {
            description: "Callback function triggered when the close button (x) is clicked.",
        },
    },
    args: {
        onClose: fn(),
        className: "max-w-[976px]",
    },
    parameters: {
        layout: "fullscreen",
        docs: {
            story: {
                inline: false,
                iframeHeight: 300,
            },
        },
    },
} satisfies Meta<typeof Dialog>;

export default meta;

type Story = StoryObj<typeof meta>;

export const StringTitle: Story = {
    args: {
        title: "Default Dialog Title",
        children: <p>This is the default body content for the dialog. You can put any JSX here.</p>,
    },
};

export const EmptyStringTitle: Story = {
    args: {
        title: "",
        children: <p>This is the default body content for the dialog. You can put any JSX here.</p>,
    },
};

export const JSXTitle: Story = {
    args: {
        title: (
            <div style={{ display: "flex", alignItems: "center", gap: "10px" }}>
                <span role="img" aria-label="Info Icon" style={{ fontSize: "24px" }}>
                    ℹ️
                </span>
                <span style={{ fontWeight: "bold", color: "var(--color-primary)" }}>Important Information</span>
            </div>
        ),
        children: (
            <p>
                This dialog uses a custom JSX element for its title, allowing for more complex header content like icons
                or styled text.
            </p>
        ),
    },
};

export const ComplexChildren: Story = {
    parameters: {
        layout: "fullscreen",
        docs: {
            story: {
                inline: false,
                iframeHeight: 600,
            },
        },
    },
    args: {
        title: "Add New Item",
        children: (
            <form
                onSubmit={(e) => {
                    e.preventDefault();
                    alert("Form Submitted!");
                }}
            >
                <div className="mb-4">
                    <label htmlFor="item-name" className="form-label block mb-2 text-[var(--color-text)]">
                        Item Name
                    </label>
                    <input
                        type="text"
                        id="item-name"
                        placeholder="Enter item name"
                        className="
                            input-field
                            w-full p-3 border border-[var(--color-gray)]
                            rounded-md bg-[var(--color-background-light)]
                        "
                    />
                </div>
                <div className="mb-4">
                    <label htmlFor="item-desc" className="form-label block mb-2 text-[var(--color-text)]">
                        Description
                    </label>
                    <textarea
                        id="item-desc"
                        rows={3}
                        placeholder="Describe the item"
                        className="
                            input-field
                            w-full p-3 border border-[var(--color-gray)]
                            rounded-md bg-[var(--color-background-light)]
                        "
                    ></textarea>
                </div>
                <div className="text-right mt-6">
                    <button type="reset" className="btn btn-outline mr-5">
                        Reset
                    </button>
                    <button type="submit" className="btn">
                        Submit
                    </button>
                </div>
            </form>
        ),
    },
};

export const LongContent: Story = {
    parameters: {
        docs: {
            story: {
                inline: false,
                iframeHeight: 700,
            },
        },
    },
    args: {
        title: "Dialog Scrollable Content",
        children: (
            <div className="overflow-y-auto">
                <p>
                    This dialog contains a lot of content to demonstrate the vertical scrolling behavior when the
                    content exceeds the maximum height defined by `max-h-[90vh]` and `overflow-y-auto`.
                </p>
                {[...Array(25)].map((_, i) => (
                    <p key={i} style={{ margin: "1rem 0", lineHeight: "1.6" }}>
                        This is paragraph number {i + 1}. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed
                        do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis
                        nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure
                        dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur
                        sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est
                        laborum.
                    </p>
                ))}
            </div>
        ),
        className: "max-h-[90vh]",
    },
};

export const CustomClasses: Story = {
    parameters: {
        docs: {
            story: {
                iframeHeight: 700,
            },
        },
    },
    args: {
        title: "Custom Styled Dialog",
        children: (
            <p>
                This dialog has custom classes applied to it. You can use this feature to style the dialog according to
                your design requirements.
            </p>
        ),
        className:
            "max-w-[500px] h-[600px] ring-4 ring-offset-4 ring-indigo-500 text-[var(--color-text-light)] rounded-lg shadow-lg",
    },
};

export const AddMovieForm: Story = {
    parameters: {
        docs: {
            story: {
                iframeHeight: 950,
            },
        },
    },
    args: {
        title: "ADD MOVIE",
        children: <MovieForm onSubmit={fn().mockName("movieFormSubmit")} />,
    },
};

export const EditMovieForm: Story = {
    parameters: {
        docs: {
            story: {
                iframeHeight: 950,
            },
        },
    },
    args: {
        title: "EDIT MOVIE",
        children: <MovieForm initialMovieInfo={mockMovieInfo} onSubmit={fn().mockName("movieFormSubmit")} />,
    },
};

export const DeleteConfirmation: Story = {
    parameters: {
        docs: {
            story: {
                iframeHeight: 500,
            },
        },
    },
    args: {
        title: "DELETE MOVIE",
        children: (
            <div className="flex flex-col">
                <p className="text-[var(--color-text)] opacity-80 mb-16 text-left text-2xl">
                    Are you sure you want to delete this movie?
                </p>
                <div className="flex justify-end">
                    <button type="button" className="btn" onClick={fn().mockName("confirmDeleteClick")}>
                        Confirm
                    </button>
                </div>
            </div>
        ),
        className: "max-w-[686px]",
    },
};
