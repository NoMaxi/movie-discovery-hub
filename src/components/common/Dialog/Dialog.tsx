import React, { ReactNode } from "react";
import { Portal } from "react-portal";
import { FocusTrap } from "focus-trap-react";

interface DialogProps {
    title: string | React.JSX.Element;
    children: ReactNode;
    onClose: () => void;
    className?: string;
}

export const Dialog = ({ title, children, onClose, className = "" }: DialogProps) => (
    <Portal>
        <FocusTrap
            focusTrapOptions={{
                clickOutsideDeactivates: false,
                initialFocus: false,
                escapeDeactivates: false,
            }}
        >
            <div
                data-testid="dialog-overlay"
                className="
                    dialog-overlay
                    fixed flex justify-center items-center z-1000 p-4
                    bg-black bg-opacity-70 inset-0
                "
            >
                <div
                    data-testid="dialog-content"
                    className={`
                        dialog-content
                        relative flex flex-col w-full max-h-[90vh] p-[60px]
                        bg-[var(--color-background)] rounded-md shadow-2xl
                        overflow-y-auto
                        ${className}
                    `}
                    onClick={(e) => e.stopPropagation()}
                >
                    {title && (
                        <div className="dialog-header flex justify-between items-start pb-[38px]">
                            {typeof title === "string" ? (
                                <h2 className="font-light text-[40px] text-[var(--color-text)] uppercase tracking-wide">
                                    {title}
                                </h2>
                            ) : (
                                <div>{title}</div>
                            )}
                        </div>
                    )}
                    <button
                        type="button"
                        data-testid="dialog-close-button"
                        onClick={onClose}
                        className="
                            dialog-close-button
                            flex items-center justify-center w-[20px] h-[20px]
                            absolute top-[30px] right-[30px]
                            text-[var(--color-gray-lighter)] hover:text-[var(--color-primary)]
                            before:content-[''] before:block before:absolute before:w-[2px] before:h-[24px]
                            before:bg-current before:rotate-45
                            after:content-[''] after:block after:absolute after:w-[2px] after:h-[24px]
                            after:bg-current after:-rotate-45
                            cursor-pointer transition-colors duration-150
                        "
                    >
                        <span className="sr-only">Close</span>
                    </button>
                    <div className={`dialog-body flex-grow ${!title ? "pt-0" : ""}`}>{children}</div>
                </div>
            </div>
        </FocusTrap>
    </Portal>
);
