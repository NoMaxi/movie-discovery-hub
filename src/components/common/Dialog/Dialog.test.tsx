import React from "react";
import { render, screen, within } from "@testing-library/react";
import userEvent, { UserEvent } from "@testing-library/user-event";
import "@testing-library/jest-dom";
import { Dialog } from "./Dialog";

jest.mock("focus-trap-react", () => ({
    FocusTrap: ({ children }: { children: React.ReactNode }) => <div>{children}</div>,
}));

interface DialogProps {
    title: string | React.JSX.Element;
    children: React.ReactNode;
    onClose: () => void;
    className?: string;
}

describe("Dialog", () => {
    const defaultProps: Omit<DialogProps, "onClose"> = {
        title: "Default Test Title",
        children: <p>Default test children content</p>,
    };
    const mockTitleText = "Dialog Title Text";
    let onCloseMock: jest.Mock;
    let user: UserEvent;

    beforeEach(() => {
        onCloseMock = jest.fn();
        user = userEvent.setup();
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    const renderComponent = (props: Partial<React.ComponentProps<typeof Dialog>> = {}) => {
        const mergedProps = { ...defaultProps, ...props };
        return render(<Dialog {...mergedProps} onClose={onCloseMock} />);
    };

    test("Should render correctly with string title and children", () => {
        const childrenTextContent = "Children Text Content";
        const mockChildren = <p>{childrenTextContent}</p>;
        const { asFragment } = renderComponent({ title: mockTitleText, children: mockChildren });

        expect(screen.getByTestId("dialog-overlay")).toBeInTheDocument();

        const dialogContent = screen.getByTestId("dialog-content");

        expect(dialogContent).toBeInTheDocument();

        const header = within(dialogContent).getByRole("heading", { name: mockTitleText, level: 2 });

        expect(header).toBeInTheDocument();

        const dialogBody = screen.getByTestId("dialog-body");

        expect(within(dialogBody).getByText(childrenTextContent)).toBeInTheDocument();
        expect(dialogBody).toHaveClass("dialog-body");
        expect(screen.getByTestId("dialog-close-button")).toBeInTheDocument();
        expect(onCloseMock).not.toHaveBeenCalled();
        expect(asFragment()).toMatchSnapshot();
    });

    test("Should render correctly with empty string title", () => {
        renderComponent({ title: "" });
        const dialogContent = screen.getByTestId("dialog-content");

        expect(within(dialogContent).queryByRole("heading")).not.toBeInTheDocument();
        expect(screen.getByTestId("dialog-close-button")).toBeInTheDocument();

        const dialogBody = screen.getByTestId("dialog-body");

        expect(dialogBody).toBeInTheDocument();
        expect(dialogBody).toHaveClass("dialog-body pt-0");
    });

    test("Should render correctly with JSX element as title", () => {
        const jsxTitle = <span data-testid="jsx-title">{mockTitleText}</span>;
        renderComponent({ title: jsxTitle });

        const dialogContent = screen.getByTestId("dialog-content");

        expect(within(dialogContent).getByTestId("jsx-title")).toBeInTheDocument();
        expect(within(dialogContent).getByText(mockTitleText)).toBeInTheDocument();
        expect(within(dialogContent).queryByRole("heading", { level: 2 })).not.toBeInTheDocument();
        expect(screen.getByTestId("dialog-body")).toBeInTheDocument();
        expect(screen.getByTestId("dialog-close-button")).toBeInTheDocument();
    });

    test("Should call 'onClose' prop when close button is clicked", async () => {
        renderComponent();
        const closeButton = screen.getByTestId("dialog-close-button");

        await user.click(closeButton);

        expect(onCloseMock).toHaveBeenCalledTimes(1);
    });

    test("Should apply custom classes to the dialog content element", () => {
        const customClass = "my-custom-dialog";
        renderComponent({ className: customClass });

        const dialogContent = screen.getByTestId("dialog-content");

        expect(dialogContent).toHaveClass("dialog-content");
        expect(dialogContent).toHaveClass(customClass);
    });

    test("Should render complex children correctly", () => {
        const complexChildren = (
            <form data-testid="my-form">
                <label htmlFor="name">Name</label>
                <input type="text" id="name" />
                <button type="submit">Submit</button>
            </form>
        );
        renderComponent({ children: complexChildren });

        const dialogContent = screen.getByTestId("dialog-content");
        const form = within(dialogContent).getByTestId("my-form");

        expect(form).toBeInTheDocument();
        expect(within(form).getByLabelText("Name")).toBeInTheDocument();
        expect(within(form).getByRole("button", { name: "Submit" })).toBeInTheDocument();
    });
});
