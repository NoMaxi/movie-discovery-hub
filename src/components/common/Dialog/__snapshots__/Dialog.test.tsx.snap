// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog Should render correctly with string title and children 1`] = `<DocumentFragment />`;
